src/mdarray/types.jl

```diff
@@ -0,0 +1,306 @@
+    end
+end
+
+mutable struct Dimension <: AbstractDimension
+    ptr::GDAL.GDALDimensionH
+    dataset::WeakRef            # AbstractDataset
+
+    function Dimension(ptr::GDAL.GDALDimensionH, dataset::WeakRef)
+        dimension = new(ptr, dataset)
+        add_child!(dataset, dimension)
+        return dimension
+    end
+end
+
+mutable struct IDimension <: AbstractDimension
+    ptr::GDAL.GDALDimensionH
+    dataset::WeakRef            # AbstractDataset
+
+    function IDimension(ptr::GDAL.GDALDimensionH, dataset::WeakRef)
+        dimension = new(ptr, dataset)
+        add_child!(dataset, dimension)
+        ptr != C_NULL && finalizer(destroy, dimension)
+        return dimension
+    end
+end
+
+################################################################################
+
+isnull(x::AbstractAttribute) = x.ptr == C_NULL
+isnull(x::AbstractDataset) = x.ptr == C_NULL
+isnull(x::AbstractDimension) = x.ptr == C_NULL
+isnull(x::AbstractEDTComponent) = x.ptr == C_NULL
+isnull(x::AbstractExtendedDataType) = x.ptr == C_NULL
+isnull(x::AbstractFeature) = x.ptr == C_NULL
+isnull(x::AbstractFeatureDefn) = x.ptr == C_NULL
+isnull(x::AbstractFeatureLayer) = x.ptr == C_NULL
+isnull(x::AbstractFieldDefn) = x.ptr == C_NULL
+isnull(x::AbstractGeomFieldDefn) = x.ptr == C_NULL
+isnull(x::AbstractGeometry) = x.ptr == C_NULL
+isnull(x::AbstractGroup) = x.ptr == C_NULL
+isnull(x::AbstractMDArray) = x.ptr == C_NULL
+isnull(x::AbstractRasterBand) = x.ptr == C_NULL
+isnull(x::AbstractSpatialRef) = x.ptr == C_NULL
+isnull(x::ColorTable) = x.ptr == C_NULL
+isnull(x::CoordTransform) = x.ptr == C_NULL
+isnull(x::Driver) = x.ptr == C_NULL
+isnull(x::Field) = x.ptr == C_NULL
+isnull(x::RasterAttrTable) = x.ptr == C_NULL
+isnull(x::StyleManager) = x.ptr == C_NULL
+isnull(x::StyleTable) = x.ptr == C_NULL
+isnull(x::StyleTool) = x.ptr == C_NULL
+
+################################################################################
+
+const VectorLike{T} = Union{AbstractVector{<:T},NTuple{<:Any,X where X<:T}}
+
+const IndexLike{D} = Union{
+    AbstractVector{<:Integer},
+    CartesianIndex{D},
+    NTuple{D,I where I<:Integer},
+}
+const RangeLike{D} = Union{
+    AbstractVector{<:AbstractRange{<:Integer}},
+    NTuple{D,R where R<:AbstractRange{<:Integer}},
+}
+
+################################################################################
+
+Base.unsafe_convert(::Type{Ptr{Cvoid}}, x::AbstractExtendedDataType) = x.ptr
+Base.unsafe_convert(::Type{Ptr{Cvoid}}, x::AbstractEDTComponent) = x.ptr
+Base.unsafe_convert(::Type{Ptr{Cvoid}}, x::AbstractGroup) = x.ptr
+Base.unsafe_convert(::Type{Ptr{Cvoid}}, x::AbstractMDArray) = x.ptr
+Base.unsafe_convert(::Type{Ptr{Cvoid}}, x::AbstractAttribute) = x.ptr
+Base.unsafe_convert(::Type{Ptr{Cvoid}}, x::AbstractDimension) = x.ptr
+
+################################################################################
+
+function destroy(datatype::AbstractExtendedDataType)::Nothing
+    datatype.ptr == C_NULL && return nothing
```

Comment:
Should we add this check to every other implementation of `destroy(...)` too? E.g. https://github.com/yeesian/ArchGDAL.jl/blob/4b40fe857503ec2ef8bc431919d8d31bcec7cba5/src/ogr/geometry.jl#L77-L96

To my knowledge, the list of such functions would be in https://github.com/yeesian/ArchGDAL.jl/blob/4b40fe857503ec2ef8bc431919d8d31bcec7cba5/src/context.jl#L185-L270